### PR TITLE
Feature/healthcheck

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

- Use go modules
- Upgrade to go version 1.13.8
- Add healtcheck with checkers:
  - kafka instance consumer
  - kafka instance complete producer
  - kafka error reporter producer
  - Dataset API
  - Neo4j
- Decoupled API and DB models, using the updated versions of dp-graph and dp-api-clients-go

### How to review

- Make sure the code changes make sense
- Make sure the tests pass
- You can try tye `/health` endpoint

### Who can review

Anyone.

### CODE REVIEW

File: cmd/dp-dimension-importer/main.go [hunk 1]
```diff
@@ -83,61 +80,68 @@ func main() {
...
 	select {
-	case err := <-instanceConsumer.Errors():
-		log.ErrorC("incoming instance kafka consumer receieved an error, attempting graceful shutdown", err, nil)
-	case err := <-instanceCompleteProducer.Errors():
-		log.ErrorC("completed instance kafka producer receieved an error, attempting graceful shutdown", err, nil)
-	case err := <-errorReporterProducer.Errors():
-		log.ErrorC("error reporter kafka producer recieved an error, attempting graceful shutdown", err, nil)
-	case err := <-healthCheckErrors:
-		log.ErrorC("healthcheck server returned an error, attempting graceful shutdown", err, nil)
+	case err := <-instanceConsumer.Channels().Errors:
```
Comment (nshumoogum) [inline cmd/dp-dimension-importer/main.go]: Should there be a separate go routine to watch error channels, we don't want to trigger an early shutdown

File: cmd/dp-dimension-importer/main.go [hunk 2]
```diff
@@ -83,61 +80,68 @@ func main() {
...
+	consumer.Close(shutdownCtx)
+	instanceConsumer.Close(shutdownCtx)
+	instanceCompleteProducer.Close(shutdownCtx)
+	graphDB.Close(shutdownCtx)
+	errorReporterProducer.Close(shutdownCtx)
 
 	cancel() // stop timer
-	log.Info("gracecful shutdown comeplete", nil)
+	log.Event(ctx, "graceful shutdown complete", log.INFO)
 	os.Exit(1)
```
Comment (nshumoogum) [inline cmd/dp-dimension-importer/main.go]: Should be os.Exit(0) <- successful exit

File: cmd/dp-dimension-importer/main.go [hunk 3]
```diff
@@ -83,61 +80,68 @@ func main() {
...
+	StopHealthCheck(shutdownCtx, &hc, httpServer)
 
-	instanceConsumer.StopListeningToConsumer(ctx)
-	consumer.Close(ctx)
-	instanceConsumer.Close(ctx)
-	instanceCompleteProducer.Close(ctx)
-	graphDB.Close(ctx)
-	errorReporterProducer.Close(ctx)
-	healthcheckServer.Close(ctx)
+	instanceConsumer.StopListeningToConsumer(shutdownCtx)
```
Comment (nshumoogum) [inline cmd/dp-dimension-importer/main.go]: I thought if we failed to instantiate any of these dependencies on start up then we won't want to attempt to close them?

Comment (nshumoogum) [reply cmd/dp-dimension-importer/main.go]: Hence the initialise package in other services

File: cmd/producer/main.go [hunk 4]
```diff
@@ -17,11 +19,12 @@ var kafkaHost = flag.String("kafka", "localhost:9092", "")
 func main() {
 
 	flag.Parse()
+	ctx := context.Background()
 
 	var brokers []string
 	brokers = append(brokers, *kafkaHost)
 
-	producer, _ := kafka.NewProducer(brokers, *topic, int(2000000))
+	producer, _ := kafka.NewProducer(ctx, brokers, *topic, int(2000000), kafka.CreateProducerChannels())
```
Comment (nshumoogum) [inline cmd/producer/main.go]: I would handle the error from initialising kafka producer

File: handler/incoming_instance_handler.go [hunk 5]
```diff
@@ -44,38 +37,39 @@ type InstanceEventHandler struct {
...
+	ctx := context.Background()
 	err := hdlr.validate(newInstance)
 	if err != nil {
 		return err
 	}
 
-	logData := log.Data{"instance_id": newInstance.InstanceID}
-	logger.Info("handling new instance event", logData)
+	logData := log.Data{"instance_id": newInstance.InstanceID, "package": packageName}
+	log.Event(ctx, "handling new instance event", logData)
```
Comment (nshumoogum) [inline handler/incoming_instance_handler.go]: Missing log level `log.INFO`

File: handler/incoming_instance_handler.go [hunk 6]
```diff
@@ -92,7 +86,7 @@ func (hdlr *InstanceEventHandler) Handle(newInstance event.NewInstance) error {
 		return errors.Wrap(err, "Producer.Completed returned an error")
 	}
 
-	logger.Info("instance processing completed successfully", log.Data{"processing_time": time.Since(start).Seconds()})
+	log.Event(context.Background(), "instance processing completed successfully", log.INFO, log.Data{"package": packageName, "processing_time": time.Since(start).Seconds()})
```
Comment (nshumoogum) [inline handler/incoming_instance_handler.go]: Don't create new context here, use `ctx`

File: handler/incoming_instance_handler.go [hunk 7]
```diff
@@ -111,30 +105,31 @@ func (hdlr *InstanceEventHandler) validate(newInstance event.NewInstance) error
 	return nil
 }
 
-func (hdlr *InstanceEventHandler) insertDimensions(instance *model.Instance, dimensions []*model.Dimension) error {
+// func (hdlr *InstanceEventHandler) insertDimensions(ctx context.Context, instance *model.Instance, dimensions []*model.Dimension) error {
```
Comment (nshumoogum) [inline handler/incoming_instance_handler.go]: Can remove commented out code

File: handler/incoming_instance_handler.go [hunk 8]
```diff
@@ -111,30 +105,31 @@ func (hdlr *InstanceEventHandler) validate(newInstance event.NewInstance) error
...
+func (hdlr *InstanceEventHandler) insertDimensions(ctx context.Context, instance *model.Instance, dimensions []*model.Dimension) error {
 
 	cache := make(map[string]string)
 
 	for _, dimension := range dimensions {
 
-		dimension, err := hdlr.Store.InsertDimension(context.Background(), cache, instance, dimension)
+		dbDimension, err := hdlr.Store.InsertDimension(context.Background(), cache, instance.DbModel().InstanceID, dimension.DbModel())
```
Comment (nshumoogum) [inline handler/incoming_instance_handler.go]: Wary of creating new context values when we should be re-using the current context passed into function

File: message/consumer.go [hunk 9]
```diff
@@ -31,41 +25,42 @@ type Consumer struct {
...
+		logData := log.Data{"package": packageName}
 		for {
 			select {
-			case consumedMessage := <-c.consumer.Incoming():
-				loggerC.Info("consumer.incoming receieved a message", nil)
-				c.messageReciever.OnMessage(consumedMessage)
+			case consumedMessage := <-c.consumer.Channels().Upstream:
+				log.Event(c.ctx, "consumer.incoming received a message", log.INFO, logData)
+				c.messageReceiver.OnMessage(consumedMessage)
 				c.consumer.Release()
```
Comment (nshumoogum) [inline message/consumer.go:61]: Should this be changed to commitAndRelease ?

File: message/producer.go [hunk 10]
```diff
@@ -35,7 +29,7 @@ func (p InstanceCompletedProducer) Completed(e event.InstanceCompleted) error {
 	if avroError != nil {
 		return errors.Wrap(avroError, fmt.Sprintf("Marshaller.Marshal returned an error: event=%v", e))
 	}
-	p.Producer.Output() <- bytes
-	loggerP.Info("completed successfully", log.Data{"event": e})
+	p.Producer.Channels().Output <- bytes
+	log.Event(context.Background(), "completed successfully", log.INFO, log.Data{"event": e, "package": "message.InstanceCompletedProducer"})
```
Comment (nshumoogum) [inline message/producer.go]: You will probably want to pass context into Completed function to be used in log line

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: cmd/dp-dimension-importer/main.go (hunk 3 above)
Comment (DavidSubiros) [reply cmd/dp-dimension-importer/main.go]: yes, thanks!

File: cmd/producer/main.go (hunk 4 above)
Comment (DavidSubiros) [reply cmd/producer/main.go]: also doing os.Exit(1) instead of panic(err)

File: message/consumer.go (hunk 9 above)
Comment (DavidSubiros) [reply message/consumer.go:61]: At the moment we commit the message as part of the `OnMessage()` only if there is no error and then we do the release here.

kafka_message_receiver.go:
```
// OnMessage unmarshal the kafka message and pass it to the InstanceEventHandler any errors are sent to the ErrorReporter
func (r KafkaMessageReceiver) OnMessage(message kafka.Message) {
	var newInstanceEvent event.NewInstance
	// This context will come from the received kafka message
	ctx := context.Background()
	logData := log.Data{"package": "message.KafkaMessageReceiver"}
	if err := schema.NewInstanceSchema.Unmarshal(message.GetData(), &newInstanceEvent); err != nil {
		log.Event(ctx, "error while attempting to unmarshal kafka message into event.NewInstance", log.ERROR, log.Error(err), logData)
		return
	}

	logData["event"] = newInstanceEvent
	log.Event(ctx, "successfully unmarshalled kafka message into event.NewInstance", log.INFO, logData)

	if err := r.InstanceHandler.Handle(ctx, newInstanceEvent); err != nil {
		log.Event(ctx, "InstanceHandler.Handle returned an error", log.ERROR, log.Error(err), logData)
		if err := r.ErrorReporter.Notify(newInstanceEvent.InstanceID, "InstanceHandler.Handle returned an unexpected error", err); err != nil {
			log.Event(ctx, "ErrorReporter.Notify returned an error", log.ERROR, log.Error(err), logData)
		}
		return
	}

	log.Event(ctx, "newInstance event successfully processed", log.INFO, logData)
	message.Commit()
}
```

**Review by nshumoogum — APPROVED** [review]

**Pull request merged**